---
title: 'Case study: OE dataset'
author: "Koen Van den Berge"
date: "`r Sys.Date()`"
---

# pca data

```{r loadPCAData}
load("~/PhD_Data/singleCell/fletcher/ShareWithKelly/E4c2b_slingshot_wsforkelly.RData")
library(slingshot)
library(rgl)
library(rafalib)
mypar()
library(RColorBrewer)
library(mgcv)
library(tradeSeq)

origData <- t(pcax$x %*% t(pcax$rotation))
colpal <- cc
rd <- X[, 1:5]
lin <- getLineages(rd, clusterLabels = clus.labels, start.clus = "1", end.clus = "4")
crv <- getCurves(lin)
plot(X[, 1:2], col = colpal[as.factor(clus.labels)], pch = 16)
lines(crv)
# 3D plot
rgl::plot3d(X[, 1:3], t = "p", col = colpal[as.factor(clus.labels)], alpha = 0.3, pch = 19, cex = 2, size = 8, xlab = "PC 1", ylab = "PC 2", zlab = "PC 3", aspect = "iso", box = FALSE, axes = FALSE)
# rgl::plot3d(X[,1:3], t='p', col=c("white","black")[hlp+1],alpha=0.3, pch = 19, cex = 2, size=8, xlab="PC 1", ylab="PC 2", zlab="PC 3", aspect="iso", box=FALSE, axes=FALSE)
rgl::axes3d(tick = FALSE)
rgl::par3d(windowRect = c(20, 30, 800, 800))
for (i in seq_along(curves)) {
  rgl::plot3d(crv@curves[[i]]$s[order(crv@curves[[i]]$lambda), 1:3], type = "l", add = TRUE, lwd = 4, col = colpal[which.max(tail(lin@lineages[[i]], 1) == levels(clus.labels))])
}
#rgl::plot3d(20, 90, 0, col = "black", add = TRUE)


# rgl.postscript("~/fletcher3d.pdf", fmt="pdf")
```

In the plot below, trajectory 1 (neuronal trajectory) is the green curve, trajectory 2 is the yellow curve, and trajectory 3 (sustentacular trajectory) is the brown curve.

```{r plot1, echo=TRUE, fig.cap="Fletcher 3D PCA", include=identical(knitr:::pandoc_to(), 'html')}
knitr::include_graphics("/Users/koenvandenberge/fletcher3d.pdf")
#![Fletcher 3D PCA](/Users/koenvandenberge/fletcher3d.pdf)
```

```{r cufflinksCountData}
suppressPackageStartupMessages(library(SummarizedExperiment))
# download object from GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE95601
load("~/Downloads/GSE95601_oeHBCdiff_Cufflinks_eSet.rda")
counts <- assayData(Cufflinks_eSet)$counts_table
counts <- counts[!apply(counts, 1, function(row) any(is.na(row))), ]
counts <- counts[, colnames(counts) %in% colnames(origData)]
keep <- rowSums(edgeR::cpm(counts) > 5) >= 15
countsFiltered <- counts[keep, ]
countsFiltered <- countsFiltered[-grep(rownames(countsFiltered), pattern="ERCC"),]
```

```{r EDA}
# batch from Github
library(scone)
library(edgeR)
load("~/p63-HBC-diff/output/clust/oeHBCdiff/oeHBCdiff_scone.Rda")
batch <- droplevels(colData(scone_out)$batch[match(colnames(countsFiltered), rownames(colData(scone_out)))])
run <- droplevels(phenoData(Cufflinks_eSet)[[3]])[colnames(exprs(Cufflinks_eSet)) %in% colnames(origData)]
# table(batch,run)
```

# fit smoothers on unaligned data

## get ZI weights from zinbwave

```{r zinbwave}
library(zinbwave)
library(BiocParallel)
library(doParallel)
NCORES <- 2
registerDoParallel(NCORES)
register(DoparParam())

core <- SummarizedExperiment(countsFiltered, colData = data.frame(clusLabel = clus.labels, batch = batch))
#zinb_c <- zinbFit(core, X = '~ clusLabel + batch', commondispersion = TRUE)
#save(zinb_c,file="~/zinbFletcherTradeSeq.rda")
load("~/zinbFletcherTradeSeq.rda")
weights <- computeObservationalWeights(zinb_c, countsFiltered)

```


```{r fitGAM}
#gamList <- tradeSeq::fitGAM(countsFiltered, U=model.matrix(~-1+batch), pseudotime=slingPseudotime(crv, na=FALSE), cellWeights=slingCurveWeights(crv), weights=weights, nknots=6)
#save(gamList,file="~/gamListOE_tradeSeq_6k.rda")
load("~/gamListOE_tradeSeq_6k.rda")
## check convergence
converged <- unlist(lapply(gamList, function(m){
  if(class(m)=="try-error"){
    return(FALSE)
  } else return(m$converged)
}))
mean(converged)
sum(converged) ; length(gamList)
```


# BETWEEN LINEAGE COMPARISONS

## Comparing the expression at the differentiated cell type

```{r}
endTestGam <- diffEndTest(gamList, global = TRUE, pairwise = TRUE)
endOmnibusPval <- endTestGam$pvalue
sum(is.na(endOmnibusPval)) # genes we could not fit or test.
endOmnibusPadj <- p.adjust(endOmnibusPval, "fdr")
sum(endOmnibusPadj <= 0.05, na.rm = TRUE)
mean(endOmnibusPadj <= 0.05, na.rm = TRUE)
deGenesEndGam <- which(endOmnibusPadj <= 0.05)
hist(endOmnibusPval)
```

# edgeR analysis

In the edgeR analysis, I compare the gene expression of the differentiated cell types (i.e. final cluster of a trajectory) to each other, between the different trajectories.

```{r}
library(edgeR)
d <- DGEList(countsFiltered)
d <- calcNormFactors(d)
design <- model.matrix(~clus.labels+batch)
d$weights <- weights
d <- estimateDisp(d,design)
fit <- glmFit(d,design)
L <- matrix(0, nrow = ncol(fit$coefficients), ncol = 3)
rownames(L) <- colnames(fit$coefficients)
# trajectory 1 vs. 2
L[c("clus.labels12", "clus.labels15"), 1] <- c(1, -1)
# trajectory 3 vs 1
L[c("clus.labels4", "clus.labels12"), 2] <- c(1, -1)
# trajectory 3 vs 2
L[c("clus.labels4", "clus.labels15"), 3] <- c(1, -1)
lrt <- zinbwave::glmWeightedF(fit, contrast = L)
deGenesEdgeR <- which(p.adjust(lrt$table$PValue, "fdr") <= 0.05)
length(deGenesEdgeR)
mean(deGenesEdgeR %in% deGenesEndGam)
hist(lrt$table$PValue)
```

# compare analyses

We retrieve 85% of the genes that edgeR finds, but also obtain a bunch of other genes.

```{r}
mypar(mfrow=c(1,1))

## compare
edgeRDE <- p.adjust(lrt$table$PValue, "fdr") <= 0.05
gamDE <- endOmnibusPadj <= 0.05
vennC <- cbind(edgeRDE, gamDE)
vennDiagram(vennC, main = "end point comparison across all trajectories")
```

# plot unique GAM genes

```{r}
uniqueGamEndId <- which(gamDE == TRUE & edgeRDE == FALSE)

i <- 0
while (i < 10) {
  i <- i + 1
  plotSmoothers(gamList[[uniqueGamEndId[i]]])
}
```

# plot unique edgeR genes

```{r}
uniqueEdgeREndId <- which(gamDE == FALSE & edgeRDE == TRUE)

i <- 0
while (i < 10) {
  i <- i + 1
  plotSmoothers(gamList[[uniqueEdgeREndId[i]]])
}
```

# are unique GAM genes relevant?

```{r}
uniqGamEndFletcher <- rownames(countsFiltered)[uniqueGamEndId]
write.table(uniqGamEndFletcher, file = "~/uniqGamEndFletcher.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)
# submit to http://software.broadinstitute.org/gsea/msigdb/compute_overlaps.jsp for top 20 gene sets
overlapUniqEndGam <- readLines("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/tradeSeqPaper/case/fletcher/overlapUniqGamEndFletcher_6k")
overlapSets <- overlapUniqEndGam[10:30]
overlapSetsSplit <- sapply(overlapSets, function(x) strsplit(x, split = "\t"))
gsNames <- unname(unlist(lapply(overlapSetsSplit, "[[", 1)))
genesInSet <- unname(unlist(lapply(overlapSetsSplit, "[[", 2)))
genesInOverlap <- unname(unlist(lapply(overlapSetsSplit, "[[", 4)))
pvalUniqGam <- unname(unlist(lapply(overlapSetsSplit, "[[", 6)))
qval <- unname(unlist(lapply(overlapSetsSplit, "[[", 7)))
gsNames <- tolower(gsNames)
gsNames <- gsub(x = gsNames, pattern = "_", replacement = " ")
gsNames[-1] <- unname(sapply(gsNames[-1], function(x) substr(x, 4, nchar(x))))
tabUniqGam <- data.frame(
  geneSet = gsNames[-1],
  overlap = as.numeric(genesInOverlap[-1]),
  genesInSet = as.numeric(genesInSet[-1]),
  qvalue = qval[-1]
)
library(xtable)
xtable(tabUniqGam)

## figure with overlap as scale
pval <- pvalUniqGam
tab <- tabUniqGam
library(scales)
genesInComp <- 1576 #nr of genes assessed by MSigDB
genesInUniverse <- 45956 #according to MSigDB
background <- genesInComp / genesInUniverse
tabIk <- tab[nrow(tab):1,]
setOverlap <- as.numeric(as.character(tabIk$overlap))/as.numeric(as.character(tabIk$genesInSet))

pal <- colorRampPalette(c("red","yellow"))(20)
hlpBar <- barplot(setOverlap, horiz=TRUE, xlim=c(0,0.2), col=alpha(pal[order(as.numeric(as.character(tabIk$qvalue)), decreasing=FALSE)],.2))
barplot(rep(background,20), horiz=TRUE, add=TRUE, col=alpha("grey",.3))
abline(v=background, col="red", lty=2, lwd=2)
text(y=hlpBar, x=setOverlap, label=paste0(as.character(tabIk$overlap),"/",as.character(tabIk$genesInSet)), pos=4, cex=1)
text(y=hlpBar, x=rep(0,20), label=as.character(tabIk$geneSet), pos=4, cex=.8, font=2)

## figure with p-value as scale
library(scales)
genesInComp <- 1576 #nr of genes assessed by MSigDB
genesInUniverse <- 45956 #according to MSigDB
background <- genesInComp / genesInUniverse
tabIk <- tab[nrow(tab):1,]
pvalScale <- rev(-log10(as.numeric(pval[-1])))

pal <- colorRampPalette(c("red","yellow"))(20)
hlpBar <- barplot(pvalScale, horiz=TRUE, xlim=c(0,66), col=alpha(pal[order(as.numeric(as.character(tabIk$qvalue)), decreasing=FALSE)],.2),
xlab="-log10 p-value")
#barplot(rep(background,20), horiz=TRUE, add=TRUE, col=alpha("grey",.3))
#abline(v=background, col="red", lty=2, lwd=2)
text(y=hlpBar, x=pvalScale, label=paste0(as.character(tabIk$overlap),"/",as.character(tabIk$genesInSet)), pos=4, cex=1)
text(y=hlpBar, x=rep(0,20), label=as.character(tabIk$geneSet), pos=4, cex=1.1, font=2)

```



```{r patternTest}
resPat <- patternTest(gamList, global = TRUE, pairwise = TRUE)
o <- order(resPat$waldStat, decreasing = TRUE)

i <- 0
while (i < 10) {
  i <- i + 1
  plotSmoothers(gamList[[o[i]]], main = i)
}

# for paper
source("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/tradeSeqPaper/case/fletcher/plotSmoothersFletcher.R")
library(scales)
png("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/plots/topGenesFletcherPattern_6k.png", width = 6, height = 7, units = "in", res = 300)
mypar(mfrow = c(3, 2))
i <- 0
while (i < 6) {
  i <- i + 1
  plotSmoothersIk(gamList[[o[i]]], main = rownames(resPat)[o][i])
}
dev.off()

# Stage-wise testing
library(stageR)
pScreen <- resPat$pvalue
names(pScreen) <- rownames(resPat)
pConfirmation <- cbind(resPat$pvalue_1vs2, resPat$pvalue_1vs3, resPat$pvalue_2vs3)
dimnames(pConfirmation) <- list(rownames(resPat), c("1v2", "1v3", "2v3"))
stageObj <- stageR(pScreen, pConfirmation, pScreenAdjusted = FALSE)
stageObj <- stageWiseAdjustment(stageObj, alpha = .05, method = "holm", allowNA = TRUE)
res <- getResults(stageObj)
colSums(res)
allCompGenes <- names(which(rowSums(res) == 4))
write.table(allCompGenes, file = "~/allCompGenes.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)
# submit these in http://software.broadinstitute.org/gsea/msigdb/compute_overlaps.jsp

overlap <- readLines("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/overlapCellCycleOE_6k")
overlapSets <- overlap[10:30]
overlapSetsSplit <- sapply(overlapSets, function(x) strsplit(x, split = "\t"))
gsNames <- unname(unlist(lapply(overlapSetsSplit, "[[", 1)))
genesInSet <- unname(unlist(lapply(overlapSetsSplit, "[[", 2)))
genesInOverlap <- unname(unlist(lapply(overlapSetsSplit, "[[", 4)))
qval <- unname(unlist(lapply(overlapSetsSplit, "[[", 7)))
pvalPatternOE <- unname(unlist(lapply(overlapSetsSplit, "[[", 6)))
gsNames <- tolower(gsNames)
gsNames <- gsub(x = gsNames, pattern = "_", replacement = " ")
gsNames[-1] <- unname(sapply(gsNames[-1], function(x) substr(x, 4, nchar(x))))
tabPatternOE <- data.frame(
  geneSet = gsNames[-1],
  overlap = genesInOverlap[-1],
  genesInSet = genesInSet[-1],
  qvalue = qval[-1]
)
 library(xtable)
 xtable(tabPatternOE)

# cluster fitted values of genes significant for pattern test and create a heatmap.
# or plot gene clusters their smoothing profiles
log(fitted(gamList[[1]]) + .1)
```


# within lineage comparisons

### cell cycle patterns within lineages: tradeSeq associationTest

```{r}
assocRes <- associationTest(gamList, global = TRUE, lineages = TRUE)
deAssoc1 <- rownames(assocRes)[p.adjust(assocRes$pvalue_1, "fdr") <= 0.05]
deAssoc1 <- deAssoc1[!is.na(deAssoc1)]
length(deAssoc1)

o = order(assocRes$waldStat_1, decreasing=TRUE)
write.table(head(assocRes[o,"waldStat_1",drop=FALSE], 2000), file="~/topGenesNeuronalLineage.txt", quote=FALSE, col.names=FALSE)
```

#### cell cycle gene set

```{r}
cellGenes <- read.table("~/Downloads/GO_term_summary_20181001_091947.txt", header = TRUE, sep = "\t", row.names = NULL)
cellGenes$MGI.Gene.Marker.ID <- as.character(cellGenes$MGI.Gene.Marker.ID)
cellGenes <- cellGenes[cellGenes$MGI.Gene.Marker.ID %in% rownames(countsFiltered), ]
# out of a total of 2889 genes present in filtered dataset
sum(cellGenes$MGI.Gene.Marker.ID %in% deAssoc1)

# make enrichment plot
# how many cell cycle genes were discovered along ordering of significance?
o <- order(assocRes$waldStat_1, decreasing = TRUE)
sumFoundCC <- sapply(1:nrow(assocRes), function(ii) {
  sum(cellGenes$MGI.Gene.Marker.ID %in% rownames(assocRes)[o[1:ii]], na.rm = TRUE)
})
# how many do we expect by chance?
sumFoundRandom <- (nrow(cellGenes) / nrow(assocRes)) * (1:nrow(assocRes))
# absolute
plot(x = 1:nrow(assocRes), y = sumFoundCC, type = "l")
lines(x = 1:nrow(assocRes), y = (2889 / nrow(assocRes)) * (1:nrow(assocRes)), type = "l", col = "steelblue")
expRandom <- (2889 / nrow(smootherStats)) * (1:nrow(smootherStats))
# relative
plot(x = 1:nrow(assocRes), y = sumFoundCC / sumFoundRandom, type = "l", ylab = "# genes found with tradeSeq / # genes found under random selection")
abline(h = 1, lty = 2, col = "red")


png("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/plots/cellCycleFletcher.png", width = 7, height = 7, units = "in", res = 200)
plot(x = 1:nrow(assocRes), y = sumFoundCC / sumFoundRandom, type = "l", ylab = "# genes found with tradeSeq / # genes found under random selection", bty = "n", xlab = "Gene list ordered according to significance", col = "darkgray", lwd = 1.5)
abline(h = 1, lty = 2, col = "red")
dev.off()
```


#### Heatmap of top genes

```{r}
o1 <- order(assocRes[deAssoc1, "waldStat_1"], decreasing = TRUE)
top200Assoc <- deAssoc1[o1[1:200]]

# heatmap for neuronal lineage.
source("/Users/koenvandenberge/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/tradeSeq/R/utils.R")
df <- .getPredictRangeDf(gamList[[1]], lineageId = 1, nPoints = 100)
y <- do.call(rbind, lapply(gamList[top200Assoc], predict, newdata = df, type = "link"))
yScaled <- t(scale(t(y)))
pst <- slingPseudotime(crv)
pst <- cbind(pst, clus.labels)
pst <- pst[!is.na(pst[, 1]), ]
pst <- pst[order(pst[, 1], decreasing = FALSE), ]

df$ct <- NA
df$cols <- NA
df$ct[df$t1 < 80] <- "HBC"
df$cols[df$t1 < 80] <- colpal[1]
# df$ct[df$t1>=80 & df$t1<115] <- expression(paste(Delta, "HBC1"))
df$ct[df$t1 >= 80 & df$t1 < 115] <- "HBC1"
df$cols[df$t1 >= 80 & df$t1 < 115] <- colpal[3]
# df$ct[df$t1>=90 & df$t1<100] <- expression(paste(Delta, "HBC2"))
df$ct[df$t1 >= 90 & df$t1 < 100] <- "HBC2"
df$cols[df$t1 >= 90 & df$t1 < 100] <- colpal[6]
df$ct[df$t1 >= 115 & df$t1 < 220] <- "GBC"
df$cols[df$t1 >= 115 & df$t1 < 220] <- colpal[17]
df$ct[df$t1 >= 220 & df$t1 < 250] <- "INP1"
df$cols[df$t1 >= 220 & df$t1 < 250] <- colpal[24]
df$ct[df$t1 >= 250 & df$t1 < 310] <- "INP2"
df$cols[df$t1 >= 250 & df$t1 < 310] <- colpal[4]
df$ct[df$t1 >= 310 & df$t1 < 325] <- "INP3"
df$cols[df$t1 >= 310 & df$t1 < 325] <- colpal[19]
df$ct[df$t1 >= 325 & df$t1 < 385] <- "iOSN"
df$cols[df$t1 >= 325 & df$t1 < 385] <- colpal[13]
df$ct[df$t1 >= 385] <- "mOSN"
df$cols[df$t1 >= 385] <- colpal[8]

library(pheatmap)
annoCol <- data.frame(celltype = df$ct)
rownames(annoCol) <- colnames(yScaled)
dfUniq <- unique(df[, c("ct", "cols")])
colhlp <- dfUniq$cols
names(colhlp) <- dfUniq$ct
annColors <- list(celltype = colhlp)
heatRaw <- pheatmap(yScaled, cluster_cols = FALSE, show_rownames = FALSE, show_colnames = FALSE, cutree_rows = 5, clustering.method = "ward.D", annotation_col = annoCol, annotation_colors = annColors, annotation_names_col = FALSE)

## reorder big clusters.
library(vegan)
set.seed(7)
origHClust <- heatRaw$tree_row
hlpClust <- function(hc, mat) {
  reorder(hc, wts = runif(200))
}
pheatmap(yScaled, cluster_cols = FALSE, show_rownames = FALSE, show_colnames = FALSE, cutree_rows = 5, clustering.method = "ward.D", annotation_col = annoCol, annotation_colors = annColors, annotation_names_col = FALSE, clustering_callback = hlpClust)
```

# Check for HBC markers

```{r}
startRes <- startVsEndTest(gamList)
top250Genes <- rownames(head(startRes[order(startRes$waldStat, decreasing = TRUE), ], 250))
write.table(top250Genes, file = "~/startTestOETop250_tradeSeq.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)
# submit to http://software.broadinstitute.org/gsea/msigdb/compute_overlaps.jsp for top 20 gene sets


overlap <- readLines("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/overlapStartTestOE_tradeSeq_6k")
overlapSets <- overlap[10:30]
overlapSetsSplit <- sapply(overlapSets, function(x) strsplit(x, split = "\t"))
gsNames <- unname(unlist(lapply(overlapSetsSplit, "[[", 1)))
genesInSet <- unname(unlist(lapply(overlapSetsSplit, "[[", 2)))
genesInOverlap <- unname(unlist(lapply(overlapSetsSplit, "[[", 4)))
pvalStart <- unname(unlist(lapply(overlapSetsSplit, "[[", 6)))
qval <- unname(unlist(lapply(overlapSetsSplit, "[[", 7)))
gsNames <- tolower(gsNames)
gsNames <- gsub(x = gsNames, pattern = "_", replacement = " ")
gsNames[-1] <- unname(sapply(gsNames[-1], function(x) substr(x, 4, nchar(x))))
tabStart <- data.frame(
  geneSet = gsNames[-1],
  overlap = genesInOverlap[-1],
  genesInSet = genesInSet[-1],
  qvalue = qval[-1]
)
# library(xtable)
# xtable(tabStart)
```

# figure of gene set enrichment

```{r}
library(scales)
rafalib::mypar(mar=c(2.5,4,1.5,1),mfrow=c(1,3))
pal <- colorRampPalette(c("red","yellow"))(20)

## startVsEndTest: top 250 genes
tabIk <- tabStart[nrow(tabStart):1,]
pvalScale <- rev(-log10(as.numeric(pvalStart[-1])))
hlpBar <- barplot(pvalScale, horiz=TRUE, xlim=c(0,22), col=alpha(pal[order(as.numeric(as.character(tabIk$qvalue)), decreasing=FALSE)],.2),
xlab="-log10 p-value")
#text(y=hlpBar, x=pvalScale, label=paste0(as.character(tabIk$overlap),"/",as.character(tabIk$genesInSet)), pos=4, cex=1)
txts <- paste0(as.character(tabIk$overlap),"/",as.character(tabIk$genesInSet))
for(ii in 1:20) mtext(txts[ii], side=2, at=hlpBar[ii,], las=1, cex=2/3, line=.2)
text(y=hlpBar, x=rep(0,20), label=as.character(tabIk$geneSet), pos=4, cex=1.1, font=2)
mtext("a", side=3, at=-2.5, font=2, cex=3/2)

##  patternTest: genes significant in all three comparisons
tabIk <- tabPatternOE[nrow(tabPatternOE):1,]
pvalScale <- rev(-log10(as.numeric(pvalPatternOE[-1])))
hlpBar <- barplot(pvalScale, horiz=TRUE, xlim=c(0,30), col=alpha(pal[order(as.numeric(as.character(tabIk$qvalue)), decreasing=FALSE)],.2),
xlab="-log10 p-value")
#text(y=hlpBar, x=pvalScale, label=paste0(as.character(tabIk$overlap),"/",as.character(tabIk$genesInSet)), pos=4, cex=1)
txts <- paste0(as.character(tabIk$overlap),"/",as.character(tabIk$genesInSet))
for(ii in 1:20) mtext(txts[ii], side=2, at=hlpBar[ii,], las=1, cex=2/3, line=.2)
text(y=hlpBar, x=rep(0,20), label=as.character(tabIk$geneSet), pos=4, cex=1.1, font=2)
mtext("b", side=3, at=-3, font=2, cex=3/2)

##  genes unique in tradeSeq diffEnd (vs edgeR)
tabIk <- tabUniqGam[nrow(tabUniqGam):1,]
pvalScale <- rev(-log10(as.numeric(pvalUniqGam[-1])))
hlpBar <- barplot(pvalScale, horiz=TRUE, xlim=c(0,66), col=alpha(pal[order(as.numeric(as.character(tabIk$qvalue)), decreasing=FALSE)],.2),
xlab="-log10 p-value")
#text(y=hlpBar, x=pvalScale, label=paste0(as.character(tabIk$overlap),"/",as.character(tabIk$genesInSet)), pos=4, cex=1)
txts <- paste0(as.character(tabIk$overlap),"/",as.character(tabIk$genesInSet))
for(ii in 1:20) mtext(txts[ii], side=2, at=hlpBar[ii,], las=1, cex=2/3, line=.2)
text(y=hlpBar, x=rep(0,20), label=as.character(tabIk$geneSet), pos=4, cex=1.1, font=2)
mtext("c", side=3, at=-9, font=2, cex=3/2)

```



#### early DE test


```{r}
early13 <- earlyDETest(gamList, knots = c(1, 3), nPoints = 50, global = TRUE, pairwise = TRUE)
library(stageR)
pScreen <- early13$pvalue
names(pScreen) <- rownames(countsFiltered)
pConfirmation <- cbind(early13$pvalue_1vs2, early13$pvalue_1vs3, early13$pvalue_2vs3)
rownames(pConfirmation) <- rownames(countsFiltered)
colnames(pConfirmation) <- c("1v2", "1v3", "2v3")
stageRObj <- stageR(pScreen = pScreen, pConfirmation = pConfirmation, pScreenAdjusted = FALSE)
stageRObj <- stageWiseAdjustment(stageRObj, method = "holm", alpha = 0.05, allowNA = TRUE)
res <- getResults(stageRObj)
sigAll13 <- names(which(rowSums(res) == 4))
# most of these genes seem to be very relevant.
oo <- order(early13$waldStat, decreasing=TRUE)
head(early13[oo,])

library(UpSetR)
resDf <- as.data.frame(res)
colnames(resDf) <- c("Global", "Neur. vs. Microv.", "Neur. vs. Sust.", "Microv. vs. Sust.")
upset(resDf[, -1], order.by = "freq", text.scale = c(1.3, 1.3, 1, 1, 2, 1))

plotSmoothersIk2 <- function(m, nPoints = 100, legendPos="topright", ...) {
  data <- m$model
  y <- data$y
  # construct time variable based on cell assignments.
  nCurves <- length(m$smooth)
  timeAll <- c()
  col <- c()
  for (jj in seq_len(nCurves)) {
    for (ii in 1:nrow(data)) {
      if (data[ii, paste0("l", jj)] == 1) {
        timeAll[ii] <- data[ii, paste0("t", jj)]
        col[ii] <- jj
      } else {
        next
      }
    }
  }

  # plot raw data
  # cols <- c("#E7298A", "#FF7F00", "#1F78B4")
  cols <- c("#FF7F00", "#1F78B4", "#E7298A")
  plot(
    x = timeAll, y = log(y + 1), col = alpha(cols[col], 2 / 3), pch = 16, cex = 2 / 3,
    ylab = "log(count + 1)", xlab = "Pseudotime", ...
  )

  # predict and plot smoothers across the range
  for (jj in seq_len(nCurves)) {
    df <- .getPredictRangeDf(m, jj, nPoints = nPoints)
    # average batch
    df$U[] <- 0
    df$U[,"batchGBC09B"] <- 1
    yhat <- predict(m, newdata = df, type = "response")
    lines(x = df[, paste0("t", jj)], y = log(yhat+1), col = cols[jj], lwd = 2)
  }
  # knots
  abline(v = gamList[[1]]$smooth[[1]]$xp[1], lty = 2, col = "black", lwd = 1.5)
  abline(v = gamList[[1]]$smooth[[1]]$xp[3], lty = 2, col = "black", lwd = 1.5)
  legend(legendPos, c("Neuronal", "Microvillous", "Sustentacular"),
    col = cols,
    lty = 1, lwd = 2, bty = "n", cex = 4 / 5
  )
}

plotSmoothersIk2(gamList[["Sox11"]])


control=gam.control()
control$maxit=1000
hlp = tradeSeq::fitGAM(countsFiltered[9600:9700,], U=model.matrix(~-1+batch, contrasts.arg=list(batch=contr.sum)), pseudotime=slingPseudotime(crv, na=FALSE), cellWeights=slingCurveWeights(crv), weights=weights[1300:1400,], nknots=6, control=control)

plotSmoothersIk2(hlp[["Sox11"]])


### look at TFs
sigAll <- rownames(res)[res[,1]==1]
# download list of TFs from http://www.tfcheckpoint.org/index.php/browse
tfCheck <- read.table("~/Downloads/export.txt", sep="\t", header=TRUE)
tfAll <- as.character(tfCheck[,1])
sigAllTFs = sigAll[toupper(sigAll) %in% tfAll]
write.table(sigAllTFs, file="~/sigAllTFs.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
# submit for GSEA
overlapTF <- readLines("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/tradeSeqPaper/case/fletcher/overlapEarlyDETFs")
overlapSets <- overlapTF[10:30]
overlapSetsSplit <- sapply(overlapSets, function(x) strsplit(x, split = "\t"))
gsNames <- unname(unlist(lapply(overlapSetsSplit, "[[", 1)))
genesInSet <- unname(unlist(lapply(overlapSetsSplit, "[[", 2)))
genesInOverlap <- unname(unlist(lapply(overlapSetsSplit, "[[", 4)))
pvalTF <- unname(unlist(lapply(overlapSetsSplit, "[[", 6)))
pvalTF <- as.numeric(pvalTF[-1])
qval <- unname(unlist(lapply(overlapSetsSplit, "[[", 7)))
gsNames <- tolower(gsNames)
gsNames <- gsub(x = gsNames, pattern = "_", replacement = " ")
gsNames[-1] <- unname(sapply(gsNames[-1], function(x) substr(x, 4, nchar(x))))
tabTF <- data.frame(
  geneSet = gsNames[-1],
  overlap = as.numeric(genesInOverlap[-1]),
  genesInSet = as.numeric(genesInSet[-1]),
  qvalue = qval[-1]
)
library(xtable)
xtable(tabTF)


plot(x=tabTF$overlap/tabTF$genesInSet, y=-log10(pvalTF), pch=16)

## plot epithelial TFs
# download epithelial cell differentiation gene set at http://software.broadinstitute.org/gsea/msigdb/geneset_page.jsp?geneSetName=GO_EPITHELIAL_CELL_DIFFERENTIATION
epiGenes <- as.character(read.table("~/Downloads/geneset.txt", skip=2)[,1])
epiSigTFs <- sigAllTFs[toupper(sigAllTFs) %in% epiGenes]
mypar(mfrow=c(2,2))
for(ii in 1:length(epiSigTFs)) plotSmoothersIk2(gamList[[epiSigTFs[ii]]])
earlyEpiTF <- early13[epiSigTFs,]
oo13 <- order(earlyEpiTF[,"waldStat"], decreasing=TRUE)

pos=c("topleft", "topright", "topright", "topleft", rep("topright",4))
mypar(mfrow=c(2,2), bty='l')
for(ii in c(1,3:5)) plotSmoothersIk2(gamList[[rownames(earlyEpiTF[oo13,]) [ii] ]], legendPos=pos[ii], ylim=c(0,10))


###### ADDITIONAL RESULTS
### cross-reference with TFs from Fletcher et al.
# these are TFs that are DE between clusters within a lineage
deTFs <- openxlsx::read.xlsx("~/Downloads/mmc4.xlsx")[,1] # supplementary table 3 from fletcher et al. paper
# which ones are also early DE between lineages?
earlyTFs <- deTFs[deTFs %in% sigAll]
# plot earlyTFs
pdf("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/plots/earlyTFs_OE.pdf", width = 9, height = 7)
rafalib::mypar(mfrow = c(4, 4), bty = "l")
for(ii in 1:length(earlyTFs)) plotSmoothersIk2(gamList[[earlyTFs[ii]]], main=earlyTFs[ii])
dev.off()

# plot significant genes
library(scales)
png("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/plots/earlyDE13SignGenesInAll.png", width = 9, height = 7, units = "in", res = 200)
rafalib::mypar(mfrow = c(3, 3), bty = "l")
for (ii in 1:9) {
  plotSmoothersIk2(gamList[[sigAll13[ii]]], ylim = c(0, 10), main = sigAll13[ii])
}
dev.off()

pdf("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/plots/earlyDE13SignGenesInAll_allGenes.pdf", width = 9, height = 7)
rafalib::mypar(mfrow = c(4, 4), bty = "l")
for (ii in 1:length(sigAll13)) {
  plotSmoothersIk2(gamList[[sigAll13[ii]]], ylim = c(0, 10), main = sigAll13[ii])
}
dev.off()

### custom background GSEA
library(biomaRt)
ensembl <- useMart("ENSEMBL_MART_ENSEMBL")
ensembl <- useDataset("mmusculus_gene_ensembl", mart=ensembl)
bm <- getBM(attributes=c("ensembl_gene_id","hsapiens_homolog_ensembl_gene"), # this is what we want to extract
            filters="mgi_symbol", # this determines the filter
            values=tfAll, # this are the values to filter for (get only the genes in our list)
            mart=ensembl)
ensemblTF <- bm[,1]
# for DAVID
write.table(ensemblTF, file="~/tfAllEnsembl.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)


```

# plot for paper v2

```{r}
plotSmoothersIk <- function(m, nPoints = 100, ...) {
  data <- m$model
  y <- data$y
  # construct time variable based on cell assignments.
  nCurves <- length(m$smooth)
  timeAll <- c()
  col <- c()
  for (jj in seq_len(nCurves)) {
    for (ii in 1:nrow(data)) {
      if (data[ii, paste0("l", jj)] == 1) {
        timeAll[ii] <- data[ii, paste0("t", jj)]
        col[ii] <- jj
      } else {
        next
      }
    }
  }

  # plot raw data
  # cols <- c("#E7298A", "#FF7F00", "#1F78B4")
  cols <- c("#FF7F00", "#1F78B4", "#E7298A")
  plot(
    x = timeAll, y = log(y + 1), col = alpha(cols[col], 2 / 3), pch = 16, cex = 2 / 3,
    ylab = "log(count + 1)", xlab = "Pseudotime", ...
  )

  # predict and plot smoothers across the range
  for (jj in seq_len(nCurves)) {
    df <- .getPredictRangeDf(m, jj, nPoints = nPoints)
    yhat <- predict(m, newdata = df, type = "response")
    lines(x = df[, paste0("t", jj)], y = log(yhat + 1), col = cols[jj], lwd = 2)
  }
  # knots
  abline(v = gamList[[1]]$smooth[[1]]$xp[2], lty = 2, col = "black", lwd = 1.5)
  abline(v = gamList[[1]]$smooth[[1]]$xp[4], lty = 2, col = "black", lwd = 1.5)
  legend("topleft", c("Neuronal", "Microvillous", "Sustentacular"),
    col = cols,
    lty = 1, lwd = 2, bty = "n", cex = 4 / 5
  )
}

## heatmap
png("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/plots/FigCaseFletcher_heatmap.png", width = 7, height = 6, units = "in", res = 200)
## reorder big clusters.
set.seed(7)
origHClust <- heatRaw$tree_row
hlpClust <- function(hc, mat) {
  stats::reorder(origHClust, wts = runif(200))
}
pheatmap(yScaled, cluster_cols = FALSE, show_rownames = FALSE, show_colnames = FALSE, cutree_rows = 5, clustering.method = "ward.D", annotation_col = annoCol, annotation_colors = annColors, annotation_names_col = FALSE, clustering_callback = hlpClust)
dev.off()

### early DE genes
library(scales)
png("~/Dropbox/PhD/Research/singleCell/trajectoryInference/trajectoryDE/plots/FigCaseFletcher_earlyDE_TFs.png", width = 7, height = 6, units = "in", res = 200)
rafalib::mypar(mfrow = c(2, 2), bty='l')
for(ii in c(1,3:5)) plotSmoothersIk2(gamList[[rownames(earlyEpiTF[oo13,]) [ii] ]], legendPos=pos[ii], ylim=c(0,10), main=rownames(earlyEpiTF[oo13,])[ii])
dev.off()

```


# Session information

```{r}
sessionInfo()
```
